db/db_impl.cc

```diff
@@ -956,17 +976,11 @@ Status DBImpl::SetDBOptions(
                                           : new_options.max_open_files - 10);
       wal_changed = mutable_db_options_.wal_bytes_per_sync !=
                     new_options.wal_bytes_per_sync;
-      if (new_options.bytes_per_sync == 0) {
-        new_options.bytes_per_sync = 1024 * 1024;
-      }
       mutable_db_options_ = new_options;
-      env_options_for_compaction_ = EnvOptions(
-          BuildDBOptions(immutable_db_options_, mutable_db_options_));
+      env_options_for_compaction_ = EnvOptions(new_db_options);
       env_options_for_compaction_ = env_->OptimizeForCompactionTableWrite(
           env_options_for_compaction_, immutable_db_options_);
       versions_->ChangeEnvOptions(mutable_db_options_);
-      env_options_for_compaction_ = env_->OptimizeForCompactionTableRead(
```

Comment:
This line seems to have been a mistake.

Reply:
This line is introduced in https://github.com/facebook/rocksdb/pull/3004
Previously OptimizeForCompactionTableRead would only be used to set env_options_for_read_
https://github.com/facebook/rocksdb/blob/e62986260f12abad62d84182d106daeb147168e7/db/compaction_job.cc#L328 which is used for reads. env_options_for_compaction_ which is for writes should only go through OptimizeForCompactionTableWrite.
env_options_for_compaction_

Reply:
env_options_for_compaction_ doesn't necessarily need to be for writes; at least it doesn't seem to be so from the name. I believe it is used in the lower layers to optimize compaction reads as well.

Reply:
I checked all the usages. The only place it is used for read is here: https://github.com/facebook/rocksdb/blob/master/db/compaction_job.cc#L646-L652
Which explicitly says that it should not be optimized for reads.

Reply:
If you see `OptimizeForCompactionTableRead`, it sets the `use_direct_reads` option. If `use_direct_reads` is set, it impacts both user and compaction reads. All the table readers opened for either user or compaction reads should be aware of whether to use O_DIRECT or not and need to open the files appropriately. 

If you follow further down the link you provided above, and see in `TableCache::GetTableReader`, a `NewRandomAccessFile` is opened with these env options. 
https://github.com/facebook/rocksdb/blob/master/db/table_cache.cc#L104

And Posix's `NewRandomAccessFile` makes ample use of `use_direct_reads`:
https://github.com/facebook/rocksdb/blob/master/env/env_posix.cc#L202

Reply:
I don't think so. Can you show the line that env_options_for_compaction_ is used for reading?

Moreover, env_options_for_compaction_ is not affected by OptimizeForCompactionTableRead in ::Open (which is the correct behavior) and suddenly changes when ::SetOptions is called.
If we think use_direct_reads should be used for writes (which I highly doubt), then it should be set in OptimizeForCompactionTableWrite.

Reply:
Discussed offline. Reverting the change for now to have it fixed in a later PR.